src/api.rs

```diff
@@ -0,0 +1,151 @@
+extern crate rocksdb;
+
+use std::{
+	pin::Pin,
+	sync::Arc,
+	task::{Context, Poll},
+};
+
+use ::futures::prelude::*;
+use chrono::{DateTime, Local};
+use hyper::{
+	header::ACCESS_CONTROL_ALLOW_ORIGIN, service::Service, Body, Method, Request, Response, Server,
+	StatusCode,
+};
+use regex::Regex;
+use rocksdb::{DBWithThreadMode, SingleThreaded};
+use tokio;
+
+use crate::rpc::{check_http, get_block_by_number, get_kate_query_proof_by_cell};
+
+// 💡 HTTP part where handles the RPC Queries
+
+//service part of hyper
+struct Handler {
+	store: Arc<DBWithThreadMode<SingleThreaded>>,
+	url: String,
+}
+
+impl Service<Request<Body>> for Handler {
+	type Error = hyper::Error;
+	type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+	type Response = Response<Body>;
+
+	fn poll_ready(&mut self, _: &mut Context) -> Poll<Result<(), Self::Error>> {
+		Poll::Ready(Ok(()))
+	}
+
+	fn call(&mut self, req: Request<Body>) -> Self::Future {
+		fn match_url_block(path: &str) -> Result<u64, String> {
+			let re = Regex::new(r"^(/v1/data/(\d{1,}))$").unwrap();
+			if let Some(caps) = re.captures(path) {
+				if let Some(block) = caps.get(2) {
+					return Ok(block.as_str().parse::<u64>().unwrap());
+				}
+			}
+			Err("no match found !".to_owned())
+		}
+
+		fn mk_response(s: String) -> Result<Response<Body>, hyper::Error> {
+			Ok(Response::builder()
+				.status(200)
+				.header(ACCESS_CONTROL_ALLOW_ORIGIN, "*")
+				.header("Content-Type", "application/json")
+				.body(Body::from(s))
+				.unwrap())
+		}
+
+		let local_tm: DateTime<Local> = Local::now();
+		log::info!(
+			"⚡️ {} | {} | {}",
+			local_tm.to_rfc2822(),
+			req.method(),
+			req.uri().path()
+		);
+
+		let _db = self.store.clone();
+		let url = self.url.clone();
+
+		Box::pin(async move {
+			let res = match req.method() {
+				&Method::GET => {
+					if let Ok(block_num) = match_url_block(req.uri().path()) {
+						//cell logic to be written here
+						let block = get_block_by_number(&url, block_num).await.unwrap();
+						let header = block.header;
+						let max_rows = header.extrinsics_root.rows;
+						let max_cols = header.extrinsics_root.cols;
+						let mut vec = Vec::new();
+						for i in 0..max_rows {
+							for j in 0..max_cols {
+								let cells = get_kate_query_proof_by_cell(&url, block_num, i, j)
```

Comment:
Looks like we are re-fetching all the relevant cells without relying on existing code which already fetches the data. Or am I missing something?

Reply:
I tried to fetch the HEADER from the DB that stores the headers and the confidence. But the problem was that I was not able to retrieve that properly like we want and I was also rushing for the product team. Anyways I will try again to fetch from the DB

Reply:
Yeah, we should def have an abstraction over the fetching logic, as it will be used in at least a couple of places...

Reply:
Fetching from rocksdb was working, But we were using `get_kate_proof` which fetches the relevent cells but, in case of no app_id is given it fetches the random cells, which is not the data that we want. That's what I thought on it